lib/sources.nix

```diff
@@ -176,4 +176,214 @@ rec {
   pathHasContext = builtins.hasContext or (lib.hasPrefix storeDir);
 
   canCleanSource = src: src ? _isLibCleanSourceWith || !(pathHasContext (toString src));
+
+  # Splits a filesystem path into its components.
+  splitPath = path: lib.splitString "/" (toString path);
+
+  # Turns a list of path components into a tree, e.g.
+  #
+  #     ["a" "b" "c1"]
+  #     ["a" "b" "c2"]
+  #     ["a" "b" "c3"]
+  #     ["a" "x"     ]
+  #
+  # becomes:
+  #
+  #     { a = { b = { c1 = null; c2 = null; c3 = null }; x = null; }; }
+  pathComponentsToTree = paths: with lib;
+    foldl (tree: path: recursiveUpdate tree (setAttrByPath path null)) {} paths;
+
+  # Returns true if and only if any prefix of the given `path` leads to a leaf
+  # (`null`) in the given `tree` (nested attrset).
+  # That is: "If we go down the tree by the given path, do we hit a leaf?"
+  #
+  # Example: For the tree `tree`
+  #
+  #     a
+  #       b-leaf
+  #       c-leaf
+  #       d
+  #         e-leaf
+  #     f-leaf
+  #
+  # represented as attrset
+  #
+  #     { a = { b = null; c = null; d = { e = null; }; }; f = null; }
+  #
+  # we have (string quotes omitted for readability):
+  #
+  #     isPrefixOfLeafPath [a]       tree == false
+  #     isPrefixOfLeafPath [x]       tree == false
+  #     isPrefixOfLeafPath [a b]     tree == true
+  #     isPrefixOfLeafPath [a b c]   tree == true
+  #     isPrefixOfLeafPath [a b c x] tree == true
+  #     isPrefixOfLeafPath [a d]     tree == false
+  isPrefixOfLeafPath = path: tree:
+    if tree == null
+      then true
+      else
+        if path == []
```

Comment:
Avoid the hadoukif
![image](https://user-images.githubusercontent.com/5954806/206297597-85ba1d81-0f02-4d64-b34d-3f835ba34e8d.png)

Usually we pile the `else if`s 

```
if cond1 then val1
else if cond2 then val2
else if cond3 then val3
else ...
```